specification.json

```diff
@@ -559,6 +594,48 @@
             "content": "The hook MUST NOT alter the `hook hints` structure.",
             "RFC 2119 keyword": "MUST NOT",
             "children": []
+        },
+        {
+            "id": "Requirement 5.1.1",
+            "machine_id": "requirement_5_1_1",
+            "content": "The provider MAY define a mechanism for signalling the occurrence of one of a set of events, including `PROVIDER_READY`, `PROVIDER_ERROR` and `PROVIDER_CONFIGURATION_CHANGED`.",
+            "RFC 2119 keyword": "MAY",
+            "children": []
+        },
+        {
+            "id": "Requirement 5.2.1",
+            "machine_id": "requirement_5_2_1",
+            "content": "The `client` MUST provide an `addHandler` function for attaching callbacks to `provider events`, which accepts event type(s) and a function to run when the associated event(s) occur.",
+            "RFC 2119 keyword": "MUST",
+            "children": []
+        },
```

Comment:
I am wondering if it would make sense to let providers also dispatch "custom" events.
E.g. a provider could dispatch an event like PROVIDER_STALE as @dabeeeenster said or maybe something else. This could make it easy for providers to communicate specific events. 

One problem with this proposal is, that either providers or users of the provider could couple to tightly to these events, even though this risk is small if they are only treated as "could occur" as it is for example with the PROVIDER_STALE event.

Reply:
> I am wondering if it would make sense to let providers also dispatch "custom" events.

It's something worth considering. As a counterpoint, we've tended to strictly define and enumerate any identifiers or symbols that application authors might write logic around. This preserves one of OpenFeature's key goals: to allow migration between providers with minimal code changes. If we introduced custom events, when a migration happened, you might have a bunch of events that are no longer relevant.

Reply:
> What we could do here is define the first set of events that we agree on (based on vendor feedback and experience). Later on, we can allow OFEPs to add additional events.

Defining the first set seems good. But OFEPs should only bring new events that are "generic enough" right?
If so this is a great thing. But it tackles a bit of a different problem than custom events would I would say.

Reply:
> > Besides, language-specific implementation could be done in a generic way, so that specific vendors can add custom events with custom states.
> 
> Is this in the case the spec would allow for custom events?

After thinking a bit deeper on the implementation aspect, I think spec can only **recommend** supported event types. 

The reason is its provider implementation that exposes the `addHandler` function. Even if we define the `addHandler` contract (inputs and outputs), a specific provider implementation could accept events beyond what the specification says as we have no control over inputs and outputs 🤔

Thinking from Go perspective, a pseudocode would look like the below,

```go
// response as specified by spec
type EventResponse struct {
	state       string  // states - we can recommend through spec on what to support 
	description string
}

// event handler 
type EventHandler func(*EventResponse)

// existing provider 
type Provider struct {
}

// spec compliant handler? 
func (p Provider) addHandler(EventHandler) {
     // provider internals 
}
```

 @toddbaert  may be you have other perspective on this

Reply:
That's true, but still to be spec compliant you can only return whatever the spec defines isn't it?
So this "recommendation" is good enough to define what means to be OpenFeature compliant I would say.

Reply:
>  but still to be spec compliant you can only return whatever the spec defines isn't it?

Yes, this is this correct. Adhering only to spec-defined event types makes the provider OpenFeature compliant

Reply:
> The reason is its provider implementation that exposes the `addHandler` function.

It doesn't, in this proposal. The provider has to expose some means of _emitting_ events. Handlers are added to the client.

In JS, this looks like an "emitter" on the provider, which is used to just signal the SDK that an event occurred. See this [example cloudbees provider](https://github.com/open-feature/playground/blob/main/packages/web-cloudbees-provider/src/lib/web-cloudbees-provider.ts#L56) (and the others linked in the description).